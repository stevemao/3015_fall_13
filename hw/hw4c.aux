\relax 
\@writefile{toc}{\contentsline {section}{\numberline {1}}{1}}
\@writefile{toc}{\contentsline {section}{\numberline {2}info, infix operators and precedence}{1}}
