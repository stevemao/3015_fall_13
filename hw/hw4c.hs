infinity = infinity

infixr 3 &&&
(&&&) :: Bool -> Bool -> Bool
x &&& y = {- replace False with your code using only True, False, if-then-else and x and y -} False 

infixr 2 |||
(|||) :: Bool -> Bool -> Bool
x ||| y = {- replace False with your code using only True, False, if-then-else and x and y -} False 


