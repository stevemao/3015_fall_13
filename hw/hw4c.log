This is pdfeTeX, Version 3.141592-1.21a-2.2 (Web2C 7.5.4) (format=pdflatex 2010.5.14)  3 SEP 2010 13:28
entering extended mode
**hw4c
(./hw4c.tex
LaTeX2e <2003/12/01>
Babel <v3.8d> and hyphenation patterns for american, french, german, ngerman, b
ahasa, basque, bulgarian, catalan, croatian, czech, danish, dutch, esperanto, e
stonian, finnish, greek, icelandic, irish, italian, latin, magyar, norsk, polis
h, portuges, romanian, russian, serbian, slovak, slovene, spanish, swedish, tur
kish, ukrainian, nohyphenation, loaded.
(/usr/share/texmf/tex/latex/base/article.cls
Document Class: article 2004/02/16 v1.4f Standard LaTeX document class
(/usr/share/texmf/tex/latex/base/size11.clo
File: size11.clo 2004/02/16 v1.4f Standard LaTeX file (size option)
)
\c@part=\count79
\c@section=\count80
\c@subsection=\count81
\c@subsubsection=\count82
\c@paragraph=\count83
\c@subparagraph=\count84
\c@figure=\count85
\c@table=\count86
\abovecaptionskip=\skip41
\belowcaptionskip=\skip42
\bibindent=\dimen102
)
(/usr/share/texmf/tex/latex/preprint/fullpage.sty
Package: fullpage 1999/02/23 1.1 (PWD)
\FP@margin=\skip43
)
\@input{header.aux}
\openout2 = `header.aux'.

(./header.tex (/usr/share/texmf/tex/latex/amsfonts/amssymb.sty
Package: amssymb 2002/01/22 v2.2d

(/usr/share/texmf/tex/latex/amsfonts/amsfonts.sty
Package: amsfonts 2001/10/25 v2.2f
\@emptytoks=\toks14
\symAMSa=\mathgroup4
\symAMSb=\mathgroup5
LaTeX Font Info:    Overwriting math alphabet `\mathfrak' in version `bold'
(Font)                  U/euf/m/n --> U/euf/b/n on input line 132.
)) (./remark.sty)
(/home/faculty/jlc/papers/tex-stuff/bussproofs.sty
Proof Tree (bussproofs) style macros. Version 0.6c.
\theLevel=\count87
\myMaxLevel=\count88
\myBoxA=\box26
\myBoxB=\box27
\myBoxC=\box28
\myBoxD=\box29
\myBoxLL=\box30
\myBoxRL=\box31
\thisAboveSkip=\dimen103
\thisBelowSkip=\dimen104
\newScoreStart=\dimen105
\newScoreEnd=\dimen106
\newCenter=\dimen107
\displace=\dimen108
\leftLowerAmt=\dimen109
\rightLowerAmt=\dimen110
\scoreHeight=\dimen111
\scoreDepth=\dimen112
)
\c@theorem=\count89
\c@conjecture=\count90
\c@corollary=\count91
\c@example=\count92
\c@fact=\count93
\c@lemma=\count94
\c@definition=\count95
\c@claim=\count96
\c@remark=\count97
\c@slogan=\count98
\c@axiom=\count99
\c@note=\count100
\c@problem=\count101
\c@exercise=\count102
) (./hw4c.aux)
\openout1 = `hw4c.aux'.

LaTeX Font Info:    Checking defaults for OML/cmm/m/it on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for T1/cmr/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for OT1/cmr/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for OMS/cmsy/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for OMX/cmex/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for U/cmr/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.

Underfull \hbox (badness 10000) in paragraph at lines 6--7

 []

[1



{/usr/share/texmf-var/fonts/map/pdftex/updmap/pdftex.map}]
LaTeX Font Info:    Try loading font information for U+msa on input line 102.

(/usr/share/texmf/tex/latex/amsfonts/umsa.fd
File: umsa.fd 2002/01/19 v2.2g AMS font definitions
)
LaTeX Font Info:    Try loading font information for U+msb on input line 102.

(/usr/share/texmf/tex/latex/amsfonts/umsb.fd
File: umsb.fd 2002/01/19 v2.2g AMS font definitions
)
Overfull \hbox (6.3037pt too wide) in paragraph at lines 115--121
[]\OT1/cmr/m/n/10.95 Declare two new in-fix op-er-a-tors im-ple-ment-ing log-i-
cal im-pli-ca-tion (\OT1/cmtt/m/n/10.95 ==>\OT1/cmr/m/n/10.95 ) and exclusive-
 []

[2] (./hw4c.aux) ) 
Here is how much of TeX's memory you used:
 909 strings out of 94500
 9306 string characters out of 1176766
 55781 words of memory out of 1000000
 4109 multiletter control sequences out of 10000+50000
 11577 words of font info for 45 fonts, out of 500000 for 2000
 580 hyphenation exceptions out of 1000
 27i,8n,26p,237b,183s stack positions out of 1500i,500n,5000p,200000b,5000s
PDF statistics:
 28 PDF objects out of 300000
 0 named destinations out of 131072
 1 words of extra memory for PDF output out of 65536
</usr/share/texmf/fonts/type1/bluesky/cm/cmr9.pfb></usr/share
/texmf/fonts/type1/bluesky/cm/cmr6.pfb></usr/share/texmf/fonts/type1/bluesky/cm
/cmr8.pfb></usr/share/texmf/fonts/type1/bluesky/cm/cmtt10.pfb></usr/share/texmf
/fonts/type1/bluesky/cm/cmbx12.pfb></usr/share/texmf/fonts/type1/bluesky/cm/cmr
10.pfb></usr/share/texmf/fonts/type1/bluesky/cm/cmbx10.pfb>
Output written on hw4c.pdf (2 pages, 54284 bytes).
