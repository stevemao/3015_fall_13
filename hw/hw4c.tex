\documentclass[11pt]{article}
\usepackage{fullpage}
\include{header}

\begin{document}
\homework{4}{7 September}

\section{}

\problem{Read Chapter 1 and 2 of the Bird text.}

\section{info, infix operators and precedence}

GHC will tell you the type, where the function is defined and, for infix
operators, the precedence of the operator.  Here are some examples.

\begin{program*}
\> *Main> :info (\&\&) \\
\> (\&\&) :: Bool -> Bool -> Bool    -- Defined in GHC.Base \\
\> infixr 3 \&\& \\
\> *Main> :info (||) \\
\> (||) :: Bool -> Bool -> Bool    -- Defined in GHC.Base \\
\> infixr 2 || \\
\> *Main> :info (.) \\
\> (.) :: (b -> c) -> (a -> b) -> a -> c   -- Defined in GHC.Base \\
\> infixr 9 . \\
\> *Main> :l infinity \\
\> :l infinity \\
\> [1 of 1] Compiling Main             ( infinity.hs, interpreted ) \\
\> Ok, modules loaded: Main. \\
\> *Main> :info infinity \\
\> :info infinity \\
\> infinity :: t   -- Defined at infinity.hs:1:0-7 \\
\> *Main>  \\
\end{program*}

Precedence level is an integer from 0 to 9 with (9 being the highest
precedence.)  Function application associates to the left and has precedence 10
so, for example {\tt{f x y + 7}} is {\tt{((f x) y) + 7}}.

Haskell allows users to declare your own infix operators and whether they
associate to the left ({\tt{infixl}}) or right ({\tt{infixr}}) or do not
associate ({\tt{infix}})).  You must also specify the precedence of the new
operator.

For example, in class I presented the definitions of conjunction ({\tt{\&\&}})
and disjunction ({\tt{||}}) as follows:

\begin{program*}
\> False \&\&  x = False \\
\> True \&\&  x = x \\
\> \\
\> False || x = x \\
\> True || x = True \\
\end{program*}
This code will load, but when you try to use it you will need to specify which
definition of {\tt{\&\&}} you mean to use.  \newpage{}For example, after loading the code
above I get the following behavior.
\begin{program*}
\> *Main> True \&\& False \\
\> True \&\& False \\
\>  \\
\> <interactive>:1:5: \\
\>     Ambiguous occurrence `\&\&' \\
\>     It could refer to either `Main.\&\&', defined at hw4c.hs:7:6 \\
\>                           or `Prelude.\&\&', imported from Prelude
\end{program*}

The error message is indicating that the name {\tt{\&\&}} is ambiguous. You
could fix this by specifying which of the two definitions of {\tt{\&\&}} you
intend.
\begin{program*}
\> *Main> True Main.\&\& False \\
\> False
\end{program*}

To avoid conflict with the built-in functions we might have written the
following code which declares two new infix operators that associated to the
right.

\begin{program*}
\> infixr 3 \&\&\& \\
\> (\&\&\&) :: Bool -> Bool -> Bool \\
\> False \&\&\& y = False \\
\> True \&\&\& y = y \\
\>  \\
\> infixr 2 ||| \\
\> (|||) :: Bool -> Bool -> Bool \\
\> False ||| y = y \\
\> True ||| y = True
\end{program*}

Now we can compute as we expect.
\begin{program*}
\> *Main> True \&\&\& False \\
\> False
\end{program*}



\begin{problem}
Rewrite\footnote{This is essentially problem 2.2.1 from Bird on page 34.} the definitions of {\tt{\&\&\&}} and {\tt{|||}} so that they behave just
like {\tt{\&\&}} and {\tt{||}} but only using {\tt{if-then-else}} and the
Boolean constants. Your definitions should look something like the following:
\begin{program*}
\> x \&\&\& y = ???  \\
\> x ||| y = ???
\end{program*}
The right side of your definition ({\tt{???}}) must only use the constructs
{\tt{True}}, {\tt{False}}, {\tt{if-then-else}}, and the variables {\tt{x}} and
{\tt{y}}.
\end{problem}

\begin{problem}
Declare two new infix operators implementing logical implication ({\tt{==>}})
and exclusive-or ({\tt{<+>}}). (If you do not know the truth table for
implication and exclusive-or, look at the expected output file.) Both operators
should associate to the right.  Implication should have precedence 0 (so that
it binds weaker than the other connectives) and exclusive-or should have the
same precedence as regular disjunction ({\tt{||}}).
\end{problem}


Check the linked expected output file to see the tests you should run.



\end{document}
% Local Variables:
% mode:latex
% comment-column:0
% comment-start: "% "
% compile-command: "pdflatex hw4c"
% fill-column:79
% End:




