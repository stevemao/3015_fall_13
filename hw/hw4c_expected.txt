*Main> :l hw4c
[1 of 1] Compiling Main             ( hw4c.hs, interpreted )
Ok, modules loaded: Main.
*Main> :l hw4c
[1 of 1] Compiling Main             ( hw4c.hs, interpreted )
Ok, modules loaded: Main.
*Main> :info &&&
(&&&) :: Bool -> Bool -> Bool   -- Defined at hw4c.hs:5:6-8
infixr 3 &&&
*Main> :info |||
(|||) :: Bool -> Bool -> Bool   -- Defined at hw4c.hs:11:6-8
infixr 2 |||
*Main> :info ==>
(==>) :: Bool -> Bool -> Bool   -- Defined at hw4c.hs:21:2-4
infixr 0 ==>
*Main> :info <+>
(<+>) :: Bool -> Bool -> Bool   -- Defined at hw4c.hs:17:2-4
infixr 2 <+>
*Main> True &&& True
True
*Main> True &&& False
False
*Main> False &&& True
False
*Main> False &&& False
False
*Main> False &&& infinity
False
*Main> True ||| True
True
*Main> True ||| False
True
*Main> False ||| True
True
*Main> False ||| False
False
*Main> True ||| infinity
True
*Main> True ==> True
True
*Main> True ==> False
False
*Main> False ==> True
True
*Main> False ==> False
True
*Main> False ==> infinity
True
*Main> True <+> True
False
*Main> True <+> False
True
*Main> False <+> True
True
*Main> False <+> False
False
*Main> False <+> True ==> False
False
*Main> True <+> True ==> infinity
True
*Main> 